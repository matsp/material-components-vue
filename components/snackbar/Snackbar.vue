<template>
  <div
    :class="classes"
    class="mdc-snackbar"
    @MDCSnackbar:closed="onClosed"
    @MDCSnackbar:closing="onClosing"
    @MDCSnackbar:opened="onOpened"
    @MDCSnackbar:opening="onOpening"
  >
    <div class="mdc-snackbar__surface">
      <div
        aria-live="polite"
        class="mdc-snackbar__label"
        role="status"
      >
        <slot />
      </div>
      <div
        v-if="hasDismiss || actionButtonText !== ''"
        class="mdc-snackbar__actions"
      >
        <button
          v-if="actionButtonText !== ''"
          class="mdc-button mdc-snackbar__action"
          type="button"
        />
        <button
          v-if="hasDismiss"
          class="mdc-icon-button mdc-snackbar__dismiss material-icons"
          title="Dismiss"
        >
          close
        </button>
      </div>
    </div>
  </div>
</template>

<script>
import { MDCSnackbar } from '@material/snackbar'

import { baseComponentMixin, themeClassMixin } from '../base'

export default {
  mixins: [baseComponentMixin, themeClassMixin],
  model: {
    prop: 'open',
    event: 'change'
  },
  props: {
    timeoutMs: {
      type: Number,
      default: 5000
    },
    closeOnEscape: {
      type: Boolean,
      default: true
    },
    open: {
      type: Boolean,
      default: false
    },
    labelText: {
      type: String,
      default: ''
    },
    actionButtonText: {
      type: String,
      default: ''
    },
    leading: {
      type: Boolean,
      default: false
    },
    stacked: {
      type: Boolean,
      default: false
    },
    hasDismiss: {
      type: Boolean,
      default: false
    }
  },
  data () {
    return {
      mdcSnackbar: undefined
    }
  },
  computed: {
    classes () {
      return {
        'mdc-snackbar--leading': this.leading,
        'mdc-snackbar--stacked': this.stacked
      }
    },
    model: {
      get () {
        return this.open
      },
      set (value) {
        this.$emit('change', value)
      }
    }
  },
  watch: {
    open () {
      if (this.open) this.mdcSnackbar.open()
    },
    'mdcSnackbar.isOpen': function () {
      this.model = this.mdcSnackbar.isOpen
    },
    actionButtonText () {
      this.mdcSnackbar.actionButtonText = this.actionButtonText
    },
    labelText () {
      this.mdcSnackbar.labelText = this.labelText
    },
    closeOnEscape () {
      this.mdcSnackbar.closeOnEscape = this.closeOnEscape
    },
    timeoutMs () {
      this.mdcSnackbar.timeoutMs = this.timeoutMs
    }
  },
  mounted () {
    this.mdcSnackbar = MDCSnackbar.attachTo(this.$el)
    this.mdcSnackbar.timeoutMs = this.timeoutMs
    this.mdcSnackbar.closeOnEscape = this.closeOnEscape
    this.mdcSnackbar.labelText = this.labelText
    this.mdcSnackbar.actionButtonText = this.actionButtonText
  },
  beforeDestroy () {
    this.mdcSnackbar.destroy()
  },
  methods: {
    onClosing (e) {
      this.$emit('closing', e.detail)
    },
    onClosed (e) {
      this.$emit('closed', e.detail)
    },
    onOpening () {
      this.$emit('opening')
    },
    onOpened () {
      this.$emit('opened')
    }
  }
}
</script>
