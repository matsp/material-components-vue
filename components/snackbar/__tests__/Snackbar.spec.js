import 'mutationobserver-shim'
import { mount } from '@vue/test-utils'
import Snackbar from '../Snackbar.vue'

describe('Snackbar', () => {
  it('should mount', () => {
    let wrapper = mount(Snackbar)
    expect(wrapper.isVueInstance()).toBeTruthy()
    expect(wrapper.vm.$data.mdcSnackbar).toBeDefined()
  })

  it('should render with no prop', () => {
    let wrapper = mount(Snackbar)
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-snackbar')
    // todo: why the following doesn't work
    // expect(wrapper.isVisible()).toBe(false)
    expect(wrapper.find('button').exists()).toBeFalsy()
  })

  it('should render as open', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        open: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.isVisible()).toBe(true)
  })

  it('should render as leading', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        leading: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-snackbar--leading')
  })

  it('should render as stacked', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        stacked: true
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.classes()).toContain('mdc-snackbar--stacked')
  })

  it('should render with labelText', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        labelText: 'label text'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.$data.mdcSnackbar.labelText).toEqual('label text')
    expect(wrapper.find('.mdc-snackbar__label').text()).toEqual('label text')
  })

  it('should render with actionButton', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        actionButtonText: 'button'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.vm.$data.mdcSnackbar.actionButtonText).toEqual('button')
    expect(wrapper.find('button').text()).toEqual('button')
  })

  it('should render with optional actionButton', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        actionButtonText: 'button'
      }
    })
    expect(wrapper.find('button').exists()).toBeTruthy()

    wrapper.setProps({ actionButtonText: '' })
    expect(wrapper.find('button').exists()).toBeFalsy()
  })

  it('should render and emit', () => {
    let wrapper = mount(Snackbar, {
      propsData: {
        actionButtonText: 'button'
      }
    })
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.props().open).toBe(false)
    wrapper.setProps({ open: true })
    expect(wrapper.emittedByOrder().map(e => e.name)).toEqual(['change', 'opening', 'change'])
    expect(wrapper.isVisible()).toBe(true)
    // todo: how can I wait until the end of opening and get the opened event?
  })
})
