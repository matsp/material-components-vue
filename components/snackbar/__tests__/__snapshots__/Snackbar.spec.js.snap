// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should render and emit 1`] = `
<div class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label"></div>
    <div class="mdc-snackbar__actions"><button type="button" class="mdc-button mdc-snackbar__action">button</button>
      <!---->
    </div>
  </div>
</div>
`;

exports[`Snackbar should render as leading 1`] = `
<div class="mdc-snackbar mdc-snackbar--leading">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label"></div>
    <!---->
  </div>
</div>
`;

exports[`Snackbar should render as open 1`] = `
<div class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label"></div>
    <!---->
  </div>
</div>
`;

exports[`Snackbar should render as stacked 1`] = `
<div class="mdc-snackbar mdc-snackbar--stacked">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label"></div>
    <!---->
  </div>
</div>
`;

exports[`Snackbar should render with actionButton 1`] = `
<div class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label"></div>
    <div class="mdc-snackbar__actions"><button type="button" class="mdc-button mdc-snackbar__action">button</button>
      <!---->
    </div>
  </div>
</div>
`;

exports[`Snackbar should render with labelText 1`] = `
<div class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label">label text</div>
    <!---->
  </div>
</div>
`;

exports[`Snackbar should render with no prop 1`] = `
<div class="mdc-snackbar">
  <div class="mdc-snackbar__surface">
    <div aria-live="polite" role="status" class="mdc-snackbar__label"></div>
    <!---->
  </div>
</div>
`;
